From 23d217c5cb250319164a373cc880c9b3e689dc79 Mon Sep 17 00:00:00 2001
From: Ayyappa Ch <ayyappa.chandolu@amd.com>
Date: Wed, 25 Jan 2017 22:44:25 +0530
Subject: [PATCH 1506/1722] drm/amdgpu: Limit GART size based on visible VRAM
 size

The GART table needs to be allocated in visible VRAM. Limit the GART
size so we use at most half the visible VRAM for the GART table.

Bug: SWDEV-90092

Change-Id: I4fdbe55b50c880ba9e1d5b4aff9a667e9a66a63f
Signed-off-by: Felix Kuehling <Felix.Kuehling@amd.com>
Signed-off-by: Ayyappa Ch <ayyappa.chandolu@amd.com>
---
 drivers/gpu/drm/amd/amdgpu/gmc_v7_0.c | 12 +++++++++---
 drivers/gpu/drm/amd/amdgpu/gmc_v8_0.c | 12 +++++++++---
 2 files changed, 18 insertions(+), 6 deletions(-)

diff --git a/drivers/gpu/drm/amd/amdgpu/gmc_v7_0.c b/drivers/gpu/drm/amd/amdgpu/gmc_v7_0.c
index 2e8ae46..899f622 100644
--- a/drivers/gpu/drm/amd/amdgpu/gmc_v7_0.c
+++ b/drivers/gpu/drm/amd/amdgpu/gmc_v7_0.c
@@ -385,17 +385,23 @@ static int gmc_v7_0_mc_init(struct amdgpu_device *adev)
 
 	/* Unless the user has overridden it, compute the GART size */
 	if (amdgpu_gart_size == -1) {
+		/* Maximum GTT size is limited by the GART table size
+		 * in visible VRAM and the address space. Use at most
+		 * half of each. */
+		uint64_t max_gtt_size = min(
+			adev->mc.visible_vram_size / 8 * PAGE_SIZE / 2,
+			1ULL << 39);
+
 		si_meminfo(&si);
 		/* Set the GART to map the largest size between either
-		 * VRAM capacity or double the available physical RAM,
-		 * but no larger than 512GB.
+		 * VRAM capacity or double the available physical RAM
 		 */
 		adev->mc.gtt_size = min(
 			max(
 				((uint64_t)si.totalram * si.mem_unit * 2),
 				adev->mc.mc_vram_size
 			),
-			1024ULL << 29
+			max_gtt_size
 		);
 
 		/* GART sizes computed from physical RAM capacity
diff --git a/drivers/gpu/drm/amd/amdgpu/gmc_v8_0.c b/drivers/gpu/drm/amd/amdgpu/gmc_v8_0.c
index 8eb1e9c..537c27d 100644
--- a/drivers/gpu/drm/amd/amdgpu/gmc_v8_0.c
+++ b/drivers/gpu/drm/amd/amdgpu/gmc_v8_0.c
@@ -472,17 +472,23 @@ static int gmc_v8_0_mc_init(struct amdgpu_device *adev)
 
 	/* Unless the user has overridden it, compute the GART size */
 	if (amdgpu_gart_size == -1) {
+		/* Maximum GTT size is limited by the GART table size
+		 * in visible VRAM and the address space. Use at most
+		 * half of each. */
+		uint64_t max_gtt_size = min(
+			adev->mc.visible_vram_size / 8 * PAGE_SIZE / 2,
+			1ULL << 39);
+
 		si_meminfo(&si);
 		/* Set the GART to map the largest size between either
-		 * VRAM capacity or double the available physical RAM,
-		 * but no larger than 512GB.
+		 * VRAM capacity or double the available physical RAM
 		 */
 		adev->mc.gtt_size = min(
 			max(
 				((uint64_t)si.totalram * si.mem_unit * 2),
 				adev->mc.mc_vram_size
 			),
-			1024ULL << 29
+			max_gtt_size
 		);
 
 		/* GART sizes computed from physical RAM capacity
-- 
2.7.4

